File: surface/src/main/cpp/gl_helper.cpp
//
// Created by 王志江 on 9/14/21.
//

#include "gl_helper.h"
#include "logger.h"

#define TAG "gl_helper"

namespace igw {

    GLuint LoadShader(const GLenum shaderType, const char *shaderSource) {
        int shader = glCreateShader(shaderType);
        if (shader != 0) {
            // Pass in the shader source.
            glShaderSource(shader, 1, &shaderSource, NULL);

            // Compile the shader.
            glCompileShader(shader);

            // Get the compilation status.
            int compileStatus = GL_TRUE;
            glGetShaderiv(shader, GL_COMPILE_STATUS, &compileStatus);

            // If the compilation failed, delete the shader.
            if (compileStatus == GL_FALSE) {
                GLint maxLength = 0;
                glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &maxLength);

                LOGE(TAG, "status is %d", compileStatus);
                GLchar *msg = new GLchar [maxLength];
                glGetShaderInfoLog(shader, maxLength, &maxLength, msg);

                LOGE(TAG, "Error compiling shader: %d: %s", shaderType, msg);

                delete[] msg;
                glDeleteShader(shader);
                shader = 0;
            }
        }
        return shader;
    }

    GLuint CreateProgram(const char *vertexSource, const char *fragmentSource) {
        GLuint vertexShader = LoadShader(GL_VERTEX_SHADER, vertexSource);
        if (vertexShader == 0) {
            return 0;
        }

        GLuint fragmentShader = LoadShader(GL_FRAGMENT_SHADER, fragmentSource);
        if (fragmentShader == 0) {
            return 0;
        }

        int program = glCreateProgram();
        if (program != 0) {
            glAttachShader(program, vertexShader);
            CheckGLError(TAG, "glAttachShader");
            glAttachShader(program, fragmentShader);
            CheckGLError(TAG, "glAttachShader");

            glLinkProgram(program);

            int linkStatus = GL_TRUE;
            glGetProgramiv(program, GL_LINK_STATUS, &linkStatus);
            if (linkStatus == GL_FALSE) {
                GLint maxLength = 0;
                glGetProgramiv(program, GL_INFO_LOG_LENGTH, &maxLength);

                LOGE(TAG, "status is %d", linkStatus);
                GLchar *msg = new GLchar [maxLength];
                glGetProgramInfoLog(program, maxLength, &maxLength, msg);

                LOGE(TAG, "Error compiling program: %d: %s", program, msg);

                glDeleteProgram(program);
                program = 0;
            }
        }

        // Mark the shaders for deletion once the program is deleted.
        glDeleteShader(vertexShader);
        glDeleteShader(fragmentShader);

        return program;
    }

    int CheckGLError(const char *tag, const char *op) {
        int error = glGetError();
        if (error != GL_NO_ERROR) {
            LOGE(tag, "%s: glError 0x%x", op, error);
        }
        return error;
    }
}

File: surface/src/main/java/io/github/wzhijiang/android/surface/GLHelper.java
package io.github.wzhijiang.android.surface;

import android.opengl.GLES20;
import android.util.Log;

public class GLHelper {

    private static final String TAG = GLHelper.class.getSimpleName();

    public static void checkGlError(String tag, String op) {
        int error;
        if ((error = GLES20.glGetError()) != GLES20.GL_NO_ERROR) {
            Log.e(tag, op + ": glError " + error);
            throw new RuntimeException(op + ": glError 0x" + Integer.toHexString(error));
        }
    }

    public static int createProgram(String vertexSource, String fragmentSource) {
        int vertexShader = loadShader(GLES20.GL_VERTEX_SHADER, vertexSource);
        if (vertexShader == 0) {
            return 0;
        }

        int fragmentShader = loadShader(GLES20.GL_FRAGMENT_SHADER, fragmentSource);
        if (fragmentShader == 0) {
            return 0;
        }
        
        int program = GLES20.glCreateProgram();
        if (program == 0) {
            Log.e(TAG, "Could not create program");
        }

        GLES20.glAttachShader(program, vertexShader);
        checkGlError(TAG, "glAttachShader");
        GLES20.glAttachShader(program, fragmentShader);
        checkGlError(TAG, "glAttachShader");

        GLES20.glLinkProgram(program);
        
        int[] linkStatus = new int[1];
        GLES20.glGetProgramiv(program, GLES20.GL_LINK_STATUS, linkStatus, 0);
        if (linkStatus[0] != GLES20.GL_TRUE) {
            Log.e(TAG, "Could not link program: ");
            Log.e(TAG, GLES20.glGetProgramInfoLog(program));
            GLES20.glDeleteProgram(program);
            program = 0;
        }
        return program;
    }

    public static int loadShader(int shaderType, String source) {
        int shader = GLES20.glCreateShader(shaderType);
        if (shader == 0) {
            Log.e(TAG, "Could not create shader " + shaderType);
        }
        GLES20.glShaderSource(shader, source);
        GLES20.glCompileShader(shader);

        int[] compiled = new int[1];
        GLES20.glGetShaderiv(shader, GLES20.GL_COMPILE_STATUS, compiled, 0);
        if (compiled[0] == 0) {
            Log.e(TAG, "Could not compile shader " + shaderType + ":");
            Log.e(TAG, " " + GLES20.glGetShaderInfoLog(shader));
            GLES20.glDeleteShader(shader);
            shader = 0;
        }
        return shader;
    }

    public static void createFramebuffer(int[] textures, int[] frameBuffers, int width, int height) {
        GLES20.glGenTextures(1, textures, 0);
        GLES20.glGenFramebuffers(1, frameBuffers, 0);

        // Set up the texture
        GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, textures[0]);
        GLHelper.checkGlError(TAG, "glBindTexture: " + textures[0]);
        GLES20.glTexImage2D(
                GLES20.GL_TEXTURE_2D, 0, GLES20.GL_RGBA, width, height, 0, GLES20.GL_RGBA,
                GLES20.GL_UNSIGNED_BYTE, null);
        GLHelper.checkGlError(TAG, "glTexImage2D");

        GLES20.glTexParameterf(
                GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_MIN_FILTER, GLES20.GL_LINEAR);
        GLES20.glTexParameterf(
                GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_MAG_FILTER, GLES20.GL_LINEAR);
        GLES20.glTexParameterf(
                GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_S, GLES20.GL_CLAMP_TO_EDGE);
        GLES20.glTexParameterf(
                GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_T, GLES20.GL_CLAMP_TO_EDGE);

        // Set up the framebuffer
        GLES20.glBindFramebuffer(GLES20.GL_FRAMEBUFFER, frameBuffers[0]);
        GLES20.glFramebufferTexture2D(GLES20.GL_FRAMEBUFFER, GLES20.GL_COLOR_ATTACHMENT0,
                GLES20.GL_TEXTURE_2D, textures[0], 0);
        GLHelper.checkGlError(TAG, "glFramebufferTexture2D");

        GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, 0);
        GLES20.glBindFramebuffer(GLES20.GL_FRAMEBUFFER, 0);
    }
}
